### PR TITLE
TST: optimize.linprog/milp: add tests for various bug reports

### LINKED ISSUE #18907: BUG: `presolve` option in `milp` causing feasible problem to be reported as infeasible.
### Describe your issue.

When setting `presolve=True` via `options` dictionary in `scipy.optimize.milp` a feasible problem is being reported as infeasible. The exact same problem is "made" feasible by simply swapping to `presolve=False`. 

Alternatively changing `c` also seems to make the problem feasible. Changing `c` should not change whether a problem is feasible or not. 

### Reproducing Code Example

```python
from scipy.optimize import milp
import numpy as np
inf= np.inf

# set up problem
c = np.array([-0.85850509, -0.82892676, -0.80026454, -0.63015535, -0.5099006 , -0.50077193, -0.4894404 , -0.47285865, -0.39867774, -0.38069646, -0.36733012, -0.36733012, -0.35820411, -0.31576141, -0.20626091, -0.12466144, -0.10679516, -0.1061887 , -0.1061887 , -0.1061887 , -0. , -0. , -0. , -0. , 0. , 0. , 0. , 0. ])

constraints=(np.array([[ 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 0., 0., 0., 0.], [ 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 1., 0., 0., 0., -25., -0., -0., -0.], [ 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., -1., 0., 0., 0., 0., 0., -1., 0., 0., 0., 2., 0., 0., 0.], [ 0., 0., 0., 0., 1., 1., 1., 1., 0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., -0., -25., -0., -0.], [ 0., 0., 0., 0., -1., -1., -1., -1., 0., -1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 2., 0., 0.], [ 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 1., 1., 0., 0., 0., 0., -0., -0., -25., -0.], [ 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., -1., -1., -1., 0., 0., 0., 0., 0., 0., 2., 0.], [ 1., 1., 1., 1., 0., 0., 0., 0., 1., 0., 1., 1., 1., 1., 0., 1., 1., 0., 0., 0., 0., 1., 1., 1., -0., -0., -0., -25.], [ -1., -1., -1., -1., 0., 0., 0., 0., -1., 0., -1., -1., -1., -1., 0., -1., -1., 0., 0., 0., 0., -1., -1., -1., 0., 0., 0., 2.]]),
 np.array([-inf, -inf, -inf, -inf, -inf, -inf, -inf, -inf, -inf]), np.array([100., 0., 0., 0., 0., 0., 0., 0., 0.]))

integrality = 1

bounds=(0, 1)

# solve problem
# attempt 1
r1 = milp(c=c, constraints=constraints, integrality=integrality, bounds=bounds, options={'presolve': True}) # res.message = "The problem is infeasible. (HiGHS Status 8: model_status is Infeasible; primal_status is b'Basic')"

# attempt 2
r2 = milp(c=c, constraints=constraints, integrality=integrality, bounds=bounds, options={'presolve': False}) # res.message = 'Optimization terminated successfully. (HiGHS Status 7: Optimal)'

# update c
c_updated = np.array([-0.85850509, -0.82892676, -0.80026454, -0.63015535, -0.5099006 , -0.50077193, -0.4894404 , -0.47285865, -0.39867774, -0.38069646, -0.36733012, -0.36733012, -0.35820411, -0.31576141, -0.20626091, -0.12466144, -0.10679516, -0.1061887 , -0.1061887 , -0.1061887 , -0.5 , -0. , -0. , -0. , 0. , 0. , 0. , 0. ])

# attempt 3
r3 = milp(c=c_updated, constraints=constraints, integrality=integrality, bounds=bounds, options={'presolve': True}) # res.message = 'Optimization terminated successfully. (HiGHS Status 7: Optimal)'

print(r1)
print(r2)
print(r3)
```

### Error message

```shell
N/A
```

### SciPy/NumPy/Python version and system information

```shell
1.9.0 1.22.4 sys.version_info(major=3, minor=9, micro=5, releaselevel='final', serial=0) lapack_mkl_info: NOT AVAILABLE openblas_lapack_info: libraries = ['openblas', 'openblas'] library_dirs = ['/usr/local/lib'] language = c define_macros = [('HAVE_CBLAS', None)] runtime_library_dirs = ['/usr/local/lib'] lapack_opt_info: libraries = ['openblas', 'openblas'] library_dirs = ['/usr/local/lib'] language = c define_macros = [('HAVE_CBLAS', None)] runtime_library_dirs = ['/usr/local/lib'] blas_mkl_info: NOT AVAILABLE blis_info: NOT AVAILABLE openblas_info: libraries = ['openblas', 'openblas'] library_dirs = ['/usr/local/lib'] language = c define_macros = [('HAVE_CBLAS', None)] runtime_library_dirs = ['/usr/local/lib'] blas_opt_info: libraries = ['openblas', 'openblas'] library_dirs = ['/usr/local/lib'] language = c define_macros = [('HAVE_CBLAS', None)] runtime_library_dirs = ['/usr/local/lib']
```

### PR DESCRIPTION
**Pull request by mdhaber**
#### Reference issue
Closes gh-18907  # already fixed by HiGHS 1.6.0, it seems
gh-20116
gh-20336

#### What does this implement/fix?
There were a few open bug reports that will may be closed after upgrading to the most recent HiGHS. In case we want regression tests, here they are. If not (since these are all upstream fixes), consider this a meta-issue of bugs for which we're looking for upstream fixes.

### CODE REVIEW

**Comment by Kai-Striega** [comment]
I'm +1 on adding these in. Having tests gives me a greater sense of security that the software we use is actually working

File: scipy/optimize/tests/test_linprog.py [hunk 1]
```diff
@@ -2423,6 +2476,20 @@ def test_semi_continuous(self):
         np.testing.assert_allclose(res.x, [0, 0, 1.5, 1])
         assert res.status == 0
 
+    def test_bug_20584(self):
+        """
+        Test that when integrality is a list of all zeros, linprog gives the
+        same result as when it is an array of all zeros / integrality=None
```
Comment (stefanv) [inline scipy/optimize/tests/test_linprog.py:2482]: Not sure what the SciPy policy is, but would be nice to have a link to the issue.

**Review by stefanv — APPROVED** [review]
LGTM

**Review by tupui — APPROVED** [review]
What is the hold up here? Looks ready, from what I read we can otherwise close the issues already.

**Comment by mdhaber** [comment]
Waiting until the new HiGHs gets in. I don't think these tests all pass on main.

**Comment by mdhaber** [comment]
Looks like tests were added in gh-21565. After CI finishes, I'll document what happened, and update the linked issues, and close the PR.

File: scipy/optimize/tests/test_milp.py [hunk 2]
```diff
@@ -387,7 +387,7 @@ def test_mip_rel_gap_passdown():
     # monotonically with the mip_rel_gap parameter.
     assert np.all(np.diff(sol_mip_gaps) < 0)
 
-@pytest.mark.xfail(reason='Upstream / Wrapper issue, see gh-20589')
+
```
Comment (mdhaber) [inline scipy/optimize/tests/test_milp.py:390]:
```suggestion
@pytest.mark.xfail(reason='Upstream / Wrapper issue, see gh-20116')
```

File: scipy/optimize/tests/test_linprog.py [hunk 3]
```diff
@@ -1863,7 +1863,7 @@ def test_complementary_slackness(self):
         # non-zero RHS
         assert np.allclose(res.ineqlin.marginals @ (b_ub - A_ub @ res.x), 0)
 
-    @pytest.mark.xfail(reason='Upstream / Wrapper issue, see gh-20589')
+
```
Comment (mdhaber) [inline scipy/optimize/tests/test_linprog.py:1866]:
```suggestion
    @pytest.mark.xfail(reason='Upstream / Wrapper issue, see gh-20589')
```

**Comment by mdhaber** [comment]
This does not close gh-20116 and gh-20336; those have been reported to the HiGHS developers.

@HaoZeke based on the previous test run, it looks like the test for gh-18907 was passing even on 32-bit platforms. If everything comes back green here, any reason why we shouldn't merge this and close gh-18907?

**Comment by mdhaber** [comment]
If I hear back otherwise, we can revert; otherwise, self-merging since this just uncomments a test that seems to be passing now.

**Pull request merged**